### PR TITLE
Add video framerate limit

### PR DESCRIPTION
**Pull request by HolographicWings**
Add a setting to limit the framerate of videos
Will fix cinematics when play over the encoded framerate
Exemple in Zelda: Tears of the Kingdom when playing over 30fps

The option is under General settings and disabled by default

[Commit 1](https://github.com/HolographicWings/yuzu/commit/acf0df69b8055edd56d1c4d4db3c7b05b0ada50a)
[Commit 2](https://github.com/HolographicWings/yuzu/commit/dd80c0e33274bc9092c8870277ccdf32c6a927b2) (small fix of the first commit)
[Fork and branch](https://github.com/HolographicWings/yuzu/tree/cinematic-fix-pr)

### CODE REVIEW

**Comment by Moonlacer** [comment]
Hey! This can work with mods too, right? Not just unlocked framerate?

**Comment by liamwhite** [comment]
Is it possible to sleep video presentation for the duration of (nextframe.pts - frame.pts) instead of having it be a configurable?

**Comment by HolographicWings** [comment]
> Hey! This can work with mods too, right? Not just unlocked framerate?

yes, any FPS mod will work with

**Comment by HolographicWings** [comment]
> Is it possible to sleep video presentation for the duration of (nextframe.pts - frame.pts) instead of having it be a configurable?

I'm not sure to understand what do you mean

**Comment by liamwhite** [comment]
If pts is not viable, then frame->duration may contain a duration value which can be used to sleep. (This is in units of the time_base, which is also a property of the frame).

**Comment by FearlessTobi** [comment]
I have to say, I'm not a big fan of this idea...
Isn't this something that should just be fixed in the mods instead?

**Comment by HolographicWings** [comment]
> If pts is not viable, then frame->duration may contain a duration value which can be used to sleep. (This is in units of the time_base, which is also a property of the frame).

i'm really not familiar, this patch took me a lot of time because i'm not a c++ guy xD

**Comment by HolographicWings** [comment]
> I have to say, I'm not a big fan of this idea... Isn't this something that should just be fixed in the mods instead?

legally impossible at my known
the mods reencode the videos to 60fps, so you have to share game files, which is not the most legal thing
if video run faster it's because the yuzu decoder don't wait to show the next frame

**Comment by Moonlacer** [comment]
> I have to say, I'm not a big fan of this idea...
> Isn't this something that should just be fixed in the mods instead?

If it works fine with all our supported codecs and is just a simple toggle, why would it be an issue?

**Comment by RyzenDew** [comment]
> I have to say, I'm not a big fan of this idea... Isn't this something that should just be fixed in the mods instead?

hey now you don't want a new hack? wow i am proud you guys normally just accept them left and right like candy.

File: src/video_core/host1x/codecs/codec.cpp [hunk 1]
```diff
@@ -278,14 +278,34 @@ void Codec::Decode() {
     }
 }
 
+std::chrono::steady_clock::time_point last_frame_time = std::chrono::steady_clock::now();
```
Comment (Kelebek1) [inline src/video_core/host1x/codecs/codec.cpp:281]: We don't want globals, you should move this into Codec. If you stop and re-start a game, this variable will not change, and you'll end up with a huge value, which will then overflow `min_frame_interval - elapsed` and you'll sleep for a negative amount of time, which may or may not be safe, I'm not sure.

File: src/video_core/host1x/codecs/codec.cpp [hunk 2]
```diff
@@ -278,14 +278,34 @@ void Codec::Decode() {
...
+        std::chrono::steady_clock::time_point now = std::chrono::steady_clock::now();
+        std::chrono::microseconds elapsed =
+            std::chrono::duration_cast<std::chrono::microseconds>(now - last_frame_time);
+        std::chrono::microseconds min_frame_interval =
+            std::chrono::microseconds(1000 / Settings::values.video_framerate.GetValue() * 1000);
+
+        if (elapsed < min_frame_interval) {
+            std::this_thread::sleep_for(min_frame_interval - elapsed);
+            now = std::chrono::steady_clock::now();
+            elapsed = std::chrono::duration_cast<std::chrono::microseconds>(now - last_frame_time);
```
Comment (Kelebek1) [inline src/video_core/host1x/codecs/codec.cpp:303]: This doesn't do anything and can be removed.

**Comment by Kelebek1** [comment]
Is it a good idea to be sleeping the GPU FIFO thread?

**Comment by FearlessTobi** [comment]
Fyi, there's now also an exefs mod that's supposed to fix the cutscene speed in totk without requiring changes to the emulator,

**Comment by HolographicWings** [comment]
Many updates about the ways to fix the cutscenes in TOTK since i opened this PR, i'll try to defend the reasons why it's still to be studied in my opinion.

Firstly, apologize my english, i'm not native.

In facts, theres two mods to fix them, they have pros and cons and the debates among players.

- the first fix being a 2.2gb mod, so some will instantly guess what it does, just a re-encoding of the videos at 60fps, this mod is not safe to share as the prerendered cinematics are copyrighted by nintendo, i'm unsure but this mod sound illegal. Plus it increase the load time of the game by 1 to 3 minutes for common peoples.
- theboy181 fix is more accurate than me and is an ExeFS patch, i dunno how he done but that lock the framerate to 30 when a cutscene runs.
But his mod have conflicts with the 30 and 60 FPS mods, they both patch the same memory address (0090E9C4 or 0090E990, depend of the version of the fix)
His mod only work when the setting Limit Speed Percent is enabled, but in half of cases this setting must be disabled to play with the DynamicFPS mod, and is completly incompatible with the new FPS++ mod.
By the way the mod is done for the 1.1.0 version of the game only, the 1.1.1 being out, the mod getting decapreted, theboy fully stopped to mod the game and will not update his mod.

Finally, both of theses mods are only for Zelda TOTK, not universal.

It's why i find usefull to implement this workaround.
About FIFO sleep, with explaination about how work your codec i could maybe try to find a better way to implement this with a more accurate method, i tried some ideas to, but all of them maked the cutscene stutter, i'm still investigating.

**Review by HolographicWings — COMMENTED** [review]
Okay, this last commit is stupid, i'm fixing it
EDIT : well, a watchdog seemed usefull to avoid freezes, but i dunno really where to implement it

**Comment by liamwhite** [comment]
I haven't forgotten about this PR - but I would still like to see it changed to sleep for each frame's actual duration instead of a fixed rate. If you don't know how to do this, I can probably take a look in the next few days.

**Comment by HolographicWings** [comment]
> I haven't forgotten about this PR - but I would still like to see it changed to sleep for each frame's actual duration instead of a fixed rate. If you don't know how to do this, I can probably take a look in the next few days.

For now it wait the time that remains to get up to the desired frametime
By default the setting is on 30fps (33.33ms), if the GetCurrentFrame method run in 5ms, it will sleep the fifo for 33.33 - 5 ms (so 28.33)
Accuracy is not perfect because the timer run only in the GetCurrentFrame method

**Pull request closed without merging**